conda/notices.py

```diff
@@ -0,0 +1,502 @@
+    cache_db_path = cache_dir.joinpath(NOTICES_CACHE_DB)
+    conn = sqlite3.connect(cache_db_path)
+    conn.execute(NOTICES_CACHE_TABLE_DDL)
+
+    yield conn
+
+    conn.commit()
+    conn.close()
+
+
+def mark_channel_notices_as_viewed(
+    conn: sqlite3.Connection, channel_notice: ChannelNotice
+) -> None:
+    """
+    Insert channel notice into our database marking it as read.
+    """
+    sql = f"INSERT INTO {NOTICES_CACHE_TABLE_NAME} (id, channel_name, viewed) VALUES (?, ?, ?)"
+    try:
+        conn.execute(sql, (channel_notice.id, channel_notice.channel_name, True))
+    except sqlite3.Error as exc:
+        logger.debug(exc)
+        logger.debug(channel_notice)
+
+
+def get_viewed_channel_notice_ids(
+    conn: sqlite3.Connection, channel_notices: Sequence[ChannelNotice]
+) -> Set[str]:
+    """
+    Return the ids of the channel notices which have already been seen.
+    """
+    chn_ntc_ids = tuple(chn.id for chn in channel_notices)
+    place_holders = ",".join("?" * len(chn_ntc_ids))
+    sql = f"SELECT id FROM notices WHERE viewed = true AND id in ({place_holders})"
+    result = conn.execute(sql, chn_ntc_ids)
+
+    return {row[0] for row in result.fetchall()}
+
+
+class TerminalStyle(Enum):
+    HEADER = "\033[95m"
+    OKBLUE = "\033[94m"
+    OKCYAN = "\033[96m"
+    OKGREEN = "\033[92m"
+    WARNING = "\033[93m"
+    GRAY = "\033[90m"
+    FAIL = "\033[91m"
+    ENDC = "\033[0m"
+    BOLD = "\033[1m"
+    UNDERLINE = "\033[4m"
+
+    def __str__(self):
+        return self.value
+
+    @classmethod
+    def get_color_from_level(cls, level: NoticeLevel) -> "TerminalStyle":
+        if level == NoticeLevel.CRITICAL:
+            return cls.FAIL
+        elif level == NoticeLevel.INFO:
+            return cls.OKGREEN
+        elif level == NoticeLevel.WARNING:
+            return cls.WARNING
+        else:
+            return cls.OKBLUE
+
+    @classmethod
+    def wrap_style(cls, text: Any, style: "TerminalStyle") -> str:
+        return f"{style}{text}{TerminalStyle.ENDC}"
+
+
+def print_notices(channel_notices: Sequence[ChannelNotice], ansi_colors: bool = True):
+    """
+    Accepts a list of channel notice responses and prints a display.
+    """
+    cur_chn = None
+
+    for ntc in channel_notices:
+        if cur_chn != ntc.channel_name:
+            print()
+            if ansi_colors:
+                channel_header = TerminalStyle.wrap_style("Channel:", TerminalStyle.BOLD)
+            else:
+                channel_header = "Channel:"
+            channel_header += f" {ntc.channel_name}"
+            print(channel_header)
+            cur_chn = ntc.channel_name
+        print_notice_message(ntc, ansi_colors=ansi_colors)
+        print()
+
+
+def print_notice_message(
+    notice: ChannelNotice, indent: str = "  ", ansi_colors: bool = True
+) -> None:
+    """
+    Prints a single channel notice
+    """
+    timestamp = "" if not notice.created_at else get_locale_timestamp(notice.created_at)
```

Comment:
Gotta love f-strings, i'd also flip this to get rid of `not`:

```suggestion
    timestamp = f"{notice.created_at:%c}" if notice.created_at else ""
```